### PR TITLE
remove should_render

### LINKED ISSUE #846: Nested context does no update value if outer context should not render
### Current Situation

ReactJS appears to support nested context.

However, in IDOM situations like the following simply don't render...

```python
    return ModalStateContext(
        HomepageStateContext(
            navbar(),
            modal(),
            sidebar(),
            viewport_loading_animation(),
            viewport(),
            backdrop(),
            value=home_state,
        ),
        value=modal_state,
    )
```

### Proposed Actions

Modify the `ContextProvider` to support nested contexts.

### PR DESCRIPTION
**Pull request by rmorshea**
Closes #846, #847 

## Checklist

Please update this checklist as you complete each item:

- [x] Tests have been included for all bug fixes or added functionality.
- [x] The `changelog.rst` has been updated with any significant changes.
- [x] GitHub Issues which may be closed by this Pull Request have been linked.

### CODE REVIEW

File: tests/test_core/test_layout.py [hunk 1]
```diff
@@ -1238,3 +1160,43 @@ def Child():
                 }
             ],
         }
+
+
+async def test_render_removed_context_consumer():
```
Comment (rmorshea) [inline tests/test_core/test_layout.py:1165]: @Archmonger if you're curious the code below reproduces that detached model error

**Comment by rmorshea** [comment]
I'm actually ok merging this without review.

File: tests/test_core/test_layout.py (hunk 1 above)
Comment (Archmonger) [reply tests/test_core/test_layout.py:1165]: Is it the nonlocals that are causing the issue?

File: docs/source/about/changelog.rst [hunk 2]
```diff
@@ -23,7 +23,18 @@ more info, see the :ref:`Contributor Guide <Creating a Changelog Entry>`.
...
+**Removed**
+
+- :pull:`870` - ``ComponentType.should_render()``. This method was implemented based on
+  reading the React/Preact source code. As it turns out though it seems like it's mostly
+  a vestige from the fact that both these libraries still support class-based
+  components. The ability for components to not render also caused several bugs.
+
+**Fixed**
+
+- :issue:`846` - Nested context does no update value if outer context should not render.
```
Comment (Archmonger) [inline docs/source/about/changelog.rst:35]: does not*

File: tests/test_core/test_layout.py (hunk 1 above)
Comment (rmorshea) [reply tests/test_core/test_layout.py:1165]: No, those are just a trick to get a handle on hook callback from outside the components. The comments in the test body hopefully explain the cause. If they aren't clear I can make a follow up PR to improve the description.

**Pull request merged**